KIPs/kip-draft_kgp6.md

```diff
@@ -0,0 +1,734 @@
+ * This contract is to mainly record the addresses which holds the treasury funds
+ * before and after rebalancing. It facilates approval and redistributing to new addresses.
+ * Core will execute the re-distribution by reading this contract.
+ */
+contract TreasuryRebalance is Ownable {
+    /**
+     *  Enums to track the status of the contract
+     */
+    enum Status {
+        Initialized,
+        Registered,
+        Approved,
+        Finalized
+    }
+
+    /**
+     * Retired struct to store retired addresses and their approver addresses
+     */
+    struct Retired {
+        address retired;
+        address[] approvers;
+    }
+
+    /**
+     * Newbie struct to newbie receiver address and their fund allocation
+     */
+    struct Newbie {
+        address newbie;
+        uint256 amount;
+    }
+
+    /**
+     * Storage
+     */
+    Retired[] public retirees; // array of the Retired struct
+    Newbie[] public newbies; // array of Newbie struct
+    Status public status; // current status of the contract
+    uint256 public rebalanceBlockNumber; // the target block number of the execution of rebalancing.
+    string public memo; // result of the treasury fund rebalance. eg : `Treasury Rebalanced Successfully`
+
+    /**
+     * Events logs
+     */
+    event ContractDeployed(
+        Status status,
+        uint256 rebalanceBlockNumber,
+        uint256 deployedBlockNumber
+    );
+    event RetiredRegistered(address retired, address[] approvers);
+    event RetiredRemoved(address retired, uint256 retiredCount);
+    event NewbieRegistered(address newbie, uint256 fundAllocation);
+    event NewbieRemoved(address newbie, uint256 newbieCount);
+    event Approved(address retired, address approver, uint256 approversCount);
+    event StatusChanged(Status status);
+    event Finalized(string memo, Status status);
+
+    /**
+     * Modifiers
+     */
+    modifier onlyAtStatus(Status _status) {
+        require(status == _status, "Not in the designated status");
+        _;
+    }
+
+    /**
+     *  Constructor
+     * @param _rebalanceBlockNumber is the target block number of the execution the rebalance in Core
+     */
+    constructor(uint256 _rebalanceBlockNumber) {
+        rebalanceBlockNumber = _rebalanceBlockNumber;
+        status = Status.Initialized;
+        emit ContractDeployed(status, _rebalanceBlockNumber, block.timestamp);
+    }
+
+    //State changing Functions
+    /**
+     * @dev registers retired details
+     * @param _retiredAddress is the address of the retired
+     */
+    function registerRetired(
+        address _retiredAddress
+    ) public onlyOwner onlyAtStatus(Status.Initialized) {
+        require(
+            !retiredExists(_retiredAddress),
+            "Retired address is already registered"
+        );
+        Retired storage retired = retirees.push();
+        retired.retired = _retiredAddress;
+        emit RetiredRegistered(retired.retired, retired.approvers);
```

Comment:
At this point, approvers will always be 0, do we need to put them in the event RetiredRegistered?

Reply:
Makes sense